text/0000-numeric-literal-types.md

```diff
@@ -0,0 +1,318 @@
+- Feature Name: numeric_literal_types
+- Start Date: 2018-07-30
+- RFC PR: _
+- Rust Issue: _
+
+# Summary
+[summary]: #summary
+
+This RFC introduces two new types: `ulit` and `flit`. These are the *numeric 
+literal types*, i.e., the type of an integer literal `42` or a float literal
+`1.0`. These types exist to give a name to literals that do not have a fixed
+size. Consider the following error:
+```
+error: int literal is too large
+ --> src/main.rs:2:32
+  |
+2 | const VEGETA_CANT_EVEN: u128 = 9_000_000_000_000_000_000_000_000_000_000_000_000_001;
+  |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+```
+This expression could instead be given the type `ulit`, which could later
+be narrowed into a "real" integer, or fed into a `const` constructor,
+like `BigInt::new()`, without having to restrict itself to `u128`. These
+types, while unsized, are capable of being coerced into integers,
+following the current literal typing rules.
+
+Introducing language-level bignums is an *non-goal*.
+This RFC lays the groundwork for custom literals, but custom literals
+themselves are *also* a non-goal.
+
+Note: this proposal is given in full generality, with a series of weakened
+subsets that might be easier to implement or stabalize. The guide-level
+explanation is written only with this full generality in mind, since I don't
+think it's too difficult to explain the weakenings. Accepting this RFC will
+probably entail picking a weakening and applying it to both explanations.
+
+# Motivation
+[motivation]: #motivation
+
+This proposal has a few motivating use cases:
+- Untyped compile-time constants, as in Go or C (via `#define`).
+- Bignum constructors.
+- Custom integer literals, à la `operator""`.
+
+The former is valuable, because it allows us to hoist several occurences
+of the same literal in different typed contexts, without having to type
+it as the largest possible numeric type and explicitly narrow, i.e.
+```rust
+let foo = my_u8() & 0b0101_0101;
+let bar = my_i32() & 0b0101_0101;
+// becomes
+const MY_MASK: ulit = 0b0101_0101;
+let foo = my_u8() & MY_MASK;
+let bar = my_i32() & MY_MASK;
+// instead of
+const MY_MASK: u128 = 0b0101_0101;
+let foo = my_u8() & (MY_MASK as u8);
+let bar = my_i32() & (MY_MASK as i32);
+```
+This can be emulated by a macro that expands to the given literal, but
+that is unergonomic, and calling `MY_MASK!()` does not make it clear
+that this is a compile-time constant (`ALL_CAPS` not withstanding).
+
+The latter two are essentially the same proposal: access to arbitrary-precission
+integers for constructing bignums (and other custom literals).
+Custom literals need to take a number as input; while
+C++, the only language with custom literals, simply takes its versions of
+`u64` and `f64` as arguments for literals, this is an unnecessary restriction
+in Rust, given that we recently stabalized the `u128` type. This problem
+cannot be neatly worked around, as far as we know.
```

Comment:
The far simpler solution for custom literals is to just *take a string*. In fact C++ supports that too.

Reply:
What type would a function-style proc macro get if it was "passed" a superlong literal like this? Just a numeric literal node containing a string?

Reply:
@lxrec you get a [`Literal`](https://doc.rust-lang.org/nightly/nightly-rustc/proc_macro/struct.Literal.html). Note that there's no public methods to extract the content besides `.to_string()`.

Reply:
I don't quite like the idea of taking *literally* `&'static str`, since any application using a size that doesn't fit in the biggest type (and thus has a `FromStr` implementation) will need to parse the literal. I'm in favor of a string-based representation, but I think it should be opaque, with methods to extract components, like the exponent of a scientific notation string.

Reply:
Why? Chopping the string into its basic components is the easiest part of the parsing process by a mile. The compiler can't help with any of the actually hard parts, and inventing a whole new (thoroughly weird) kind of primitive type just to save a few library types the trouble of doing `.split('.')` and `.split('e')` seems disproportionate.